peerdroid/src/main/java/rdx/works/peerdroid/data/PeerdroidLink.kt

```diff
@@ -248,17 +241,23 @@ internal class PeerdroidLinkImpl(
 
     private suspend fun addRemoteIceCandidateInWebRtc(iceCandidate: SignalingServerMessage.RemoteData.IceCandidate) {
         val remoteIceCandidate = iceCandidate.remoteIceCandidate
-        Timber.d("🛠️️ set remote ice candidate in local WebRTC")
         webRtcManager.addRemoteIceCandidate(remoteIceCandidate = remoteIceCandidate)
     }
 
     private suspend fun terminateWithError() {
-        terminate()
+        webSocketClientJob?.cancel()
+        webSocketClient.closeSession()
+        webRtcManagerJob?.cancel()
+        webRtcManager.close()
         addConnectionDeferred.complete(Result.failure(Throwable("data channel couldn't initialize")))
     }
 
     private suspend fun terminate() {
-        Timber.d("🛠️️ terminate webrtc and web socket connection")
+        Timber.d("🗼️ terminate webrtc and web socket connection \uD83D\uDEAB")
+        webSocketClientJob?.cancel()
         webSocketClient.closeSession()
+        webRtcManagerJob?.cancel()
+        webRtcManager.close()
+        addConnectionDeferred.complete(Result.success(Unit))
```

Comment:
could you please explain what this addConnectionDeferred is responsible for?

Reply:
oh yes, in order to ensure that a new link connection has been successfully **established** we execute the following flow: 
1. scan the QR code with your wallet
2. get the connection id and start the WebRTC flow between wallet and CE
3. once the data channel is open then the CE terminates it. This is the "flag" that says the connection was successfully established
4. at this point the wallet gets notified in the `PeerConnectionEvent.Disconnected` event (previously was connected) and terminates the webrtc (the jobs and closes the streams) 

So now if notice the PeerdroidLink exposes only one function the suspend `addConnection` and returns a `Result`.
But to return the result of the above flow, I need to defer the `addConnection` fun until the whole flow is executed. And it finishes at the step 4. So there I have to complete the `addConnectionDeferred`. And then the result will be returned to the `AddLinkConnectorViewModel` which has called the `addConnection`.

Here at step 4 is what I missed before. I didn't terminate the websocket/webrtc in a proper way so that caused some issues (especially if you stay in that screen and execute transactions).

Reply:
I moved this `addConnectionDeferred.complete(Result.success(Unit))` from `PeerConnectionEvent.Disconnected -> {`